packages/buidler-core/src/internal/buidler-evm/jsonrpc/handler.ts

```diff
@@ -0,0 +1,216 @@
+import chalk from "chalk";
+import debug from "debug";
+import { IncomingMessage, ServerResponse } from "http";
+import getRawBody from "raw-body";
+import WebSocket from "ws";
+
+import { EthereumProvider } from "../../../types";
+import { BuidlerError } from "../../core/errors";
+import { ERRORS } from "../../core/errors-list";
+import {
+  isSuccessfulJsonResponse,
+  isValidJsonRequest,
+  isValidJsonResponse,
+  JsonRpcRequest,
+  JsonRpcResponse
+} from "../../util/jsonrpc";
+import {
+  BuidlerEVMProviderError,
+  InternalError,
+  InvalidJsonInputError,
+  InvalidRequestError
+} from "../provider/errors";
+
+const log = debug("buidler:core:buidler-evm:jsonrpc");
+
+export default class JsonRpcHandler {
+  private _provider: EthereumProvider;
+
+  constructor(provider: EthereumProvider) {
+    this._provider = provider;
+  }
+
+  public handleHttp = async (req: IncomingMessage, res: ServerResponse) => {
+    let rpcReq: JsonRpcRequest | undefined;
+    let rpcResp: JsonRpcResponse | undefined;
+
+    try {
+      rpcReq = await _readHttpRequest(req);
+
+      rpcResp = await this._handleRequest(rpcReq);
+    } catch (error) {
+      rpcResp = await _handleError(error);
+    }
+
+    // Validate the RPC response.
+    if (!isValidJsonResponse(rpcResp)) {
+      // Malformed response coming from the provider, report to user as an internal error.
+      rpcResp = await _handleError(new InternalError("Internal error"));
+    }
+
+    if (rpcReq !== undefined) {
+      rpcResp.id = rpcReq.id;
+    }
+
+    res.statusCode = 200;
+    res.setHeader("Content-Type", "application/json");
+    res.end(JSON.stringify(rpcResp));
+  };
+
+  public handleWs = async (ws: WebSocket) => {
+    const subscriptions: string[] = [];
+    let isClosed = false;
+
+    ws.on("message", async msg => {
+      let rpcReq: JsonRpcRequest | undefined;
+      let rpcResp: JsonRpcResponse | undefined;
+
+      try {
+        rpcReq = await _readWsRequest(msg as string);
+
+        // In case of eth_subscribe, we want to provide an emit function to the BuidlerEVM Provider.
+        if (rpcReq.method === "eth_subscribe") {
+          rpcReq.params.push(function emit(event: any) {
```

Comment:
I don't think `eth_subscribe` accepts a callback. It's a language agnostic rpc method, so there's no way of passing them.

The provider interface is still unstable, but the most common way that I know of emitting subscription events now is with a `notification` event in the provider. For example: https://github.com/xf00f/web3x/blob/master/web3x/src/providers/legacy-provider-adapter.ts#L58

Reply:
Implemented the listener via EventEmitter.

Reply:
Let me know if you find any further improvements.